pkg/cloud/cloud.go

```diff
@@ -1,37 +1,296 @@
+type SnapshotOptions struct {
+	SnapshotName   string
+	SourceVolumeId string
+	Tags           string
+}
+
+// FSx abstracts FSx client to facilitate its mocking.
+// See https://docs.aws.amazon.com/sdk-for-go/api/service/fsx/ for details
+type FSx interface {
+	CreateSnapshotWithContext(aws.Context, *fsx.CreateSnapshotInput, ...request.Option) (*fsx.CreateSnapshotOutput, error)
+	DeleteSnapshotWithContext(aws.Context, *fsx.DeleteSnapshotInput, ...request.Option) (*fsx.DeleteSnapshotOutput, error)
+	DescribeSnapshotsWithContext(aws.Context, *fsx.DescribeSnapshotsInput, ...request.Option) (*fsx.DescribeSnapshotsOutput, error)
+	DescribeFileSystemsWithContext(aws.Context, *fsx.DescribeFileSystemsInput, ...request.Option) (*fsx.DescribeFileSystemsOutput, error)
+}
+
 type Cloud interface {
-	GetMetadata() MetadataService
+	CreateSnapshot(ctx context.Context, snapshotOptions *SnapshotOptions) (snapshot *Snapshot, err error)
+	WaitForSnapshotAvailable(ctx context.Context, snapshotId string) error
+	DeleteSnapshot(ctx context.Context, snapshotId string) error
 }
 
 type cloud struct {
-	metadata MetadataService
+	fsx FSx
 }
 
 // NewCloud returns a new instance of AWS cloud
 // It panics if session is invalid
-func NewCloud() (Cloud, error) {
-	sess := session.Must(session.NewSession(&aws.Config{}))
-	svc := ec2metadata.New(sess)
+func NewCloud(region string) Cloud {
+	awsConfig := &aws.Config{
+		Region:                        aws.String(region),
+		CredentialsChainVerboseErrors: aws.Bool(true),
+	}
+
+	return &cloud{
+		fsx: fsx.New(session.Must(session.NewSession(awsConfig))),
+	}
+}
+
+func (c *cloud) CreateSnapshot(ctx context.Context, snapshotOptions *SnapshotOptions) (snapshot *Snapshot, err error) {
+	snapshotName := snapshotOptions.SnapshotName
+	volumeId := &snapshotOptions.SourceVolumeId
+	tags := snapshotOptions.Tags
+	if len(snapshotName) == 0 {
+		return nil, fmt.Errorf("snapshot name not provided")
+	}
+	if volumeId == nil {
+		return nil, fmt.Errorf("volume id not provided")
+	}
 
-	metadata, err := NewMetadataService(svc)
+	// The volume id associated with the persistent volume could refer to the file system id. In that case,
+	// we need to retrieve the id of the root volume associated with the file system
+	volumeId, err = c.getVolumeId(ctx, *volumeId)
 	if err != nil {
-		return nil, fmt.Errorf("could not get metadata from AWS: %v", err)
+		return nil, err
 	}
 
-	return &cloud{
-		metadata: metadata,
+	snapshotTags := []*fsx.Tag{
+		{
+			Key:   aws.String(SnapshotNameTagKey),
+			Value: aws.String(snapshotName),
+		},
+		{
+			Key:   aws.String(AwsFsxOpenZfsDriverTagKey),
+			Value: aws.String("true"),
+		},
+	}
+
+	if len(tags) != 0 {
+		snapshotTags = append(snapshotTags, util.ParseTags(tags)...)
+	}
+
+	input := &fsx.CreateSnapshotInput{
+		ClientRequestToken: aws.String(snapshotName),
+		Name:               aws.String(snapshotName),
+		Tags:               snapshotTags,
+		VolumeId:           volumeId,
+	}
```

Comment:
Can you add `input.Validate()`, and verify that `err == nil`? Input requirements are probably also verified server-side but this will remove the network call, decreasing execution time and reducing the load on our servers since I believe the driver retries endlessly.

Reply:
Cool, didn't know about ```.Validate()```. I"ll add this where I can.